### PR TITLE
New Package - Phase2 L1CaloTrigger

### PR DESCRIPTION
**Pull request by rekovic**
New Package for L1T Phase2: Only Standalone EG in Barrel and EndCap, 

#### PR description:

 - addressed comments,  with data files removed.
 - depends on externals cms-data/L1Trigger-L1TCalorimeter#22
 - supersedes https://github.com/cms-sw/cmssw/pull/29904

### CODE REVIEW

**Comment by rekovic** [comment]
assign upgrade

**Comment by rekovic** [comment]
assign l1

**Comment by rekovic** [comment]
test parameters

 * pull_request = cms-data/L1Trigger-L1TCalorimeter#22

**Comment by rekovic** [comment]
please test

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  edm::Handle<l1t::HGCalMulticlusterBxCollection> multiclusters_h;
+  iEvent.getByToken(multiclusters_token_, multiclusters_h);
+  const l1t::HGCalMulticlusterBxCollection &multiclusters = *multiclusters_h;
+
+  std::vector<const l1t::HGCalMulticluster *> selected_multiclusters;
+  std::map<std::pair<int, int>, std::vector<const l1t::HGCalMulticluster *>> etaphi_bins;
+
+  // here we loop on the TPGs
+  for (auto cl3d = multiclusters.begin(0); cl3d != multiclusters.end(0); cl3d++) {
+    // std::cout << "-- CL3D is EG: " <<  cl3d->hwQual() << "   "<< cl3d->eta() <<"   "<< cl3d->pt()<<std::endl;
```
Comment (rekovic) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:71]: lets leave this commented code as is, useful for checks and validation.
Will clean up in future PRs.

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            etaphi_bins[eta_phi_bin] = vec;
+          } else {
+            bucket->second.push_back(&(*cl3d));
+          }
+        }
+      }
+    }
+  }
+
+  // std::cout << "# of selected HGCAL multiclusters: " << selected_multiclusters.size() << std::endl;
```
Comment (rekovic) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:110]: lets leave this commented code as is, useful for checks and validation.
Will clean up in future PRs.

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+      reco::Candidate::PolarLorentzVector mom(pt, cl3d->eta(), cl3d->phi(), 0.);
+      reco::Candidate::PolarLorentzVector mom_eint(
+          cl3d->iPt(l1t::HGCalMulticluster::EnergyInterpretation::EM), cl3d->eta(), cl3d->phi(), 0.);
+
+      // cl3d->pt()/(1+cl3d->hOverE())
+
+      // this is not yet used
+      used_clusters.insert(cl3d);
+      auto eta_phi_bin = get_eta_phi_bin(cl3d);
+      // std::cout << " cl pt: " << cl3d->pt()
```
Comment (rekovic) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:130]: lets leave this commented code as is, useful for checks and validation.
Will clean up in future PRs.

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      for (int eta_bin : {eta_phi_bin.first - 1, eta_phi_bin.first, eta_phi_bin.first + 1}) {
+        for (int phi_bin : {eta_phi_bin.second - 1, eta_phi_bin.second, eta_phi_bin.second + 1}) {
+          auto bucket = etaphi_bins.find(std::make_pair(eta_bin, phi_bin));
+          if (bucket != etaphi_bins.end()) {
+            // this bucket is not empty
+            for (auto other_cl_ptr : bucket->second) {
+              if (used_clusters.find(other_cl_ptr) == used_clusters.end()) {
+                if (std::abs(other_cl_ptr->eta() - cl3d->eta()) < 0.02) {
+                  if (std::abs(reco::deltaPhi(other_cl_ptr->phi(), cl3d->phi())) < 0.1) {
+                    // std::cout << "MERGE with cl pt: " << other_cl_ptr->pt()
```
Comment (rekovic) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:145]: lets leave this commented code as is, useful for checks and validation.
Will clean up in future PRs.

File: L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  return std::distance(container.begin(), bin_l) - 1;
+}
+
+float L1EGammaEECalibrator::calibrationFactor(const float& pt, const float& eta) const {
+  int bin_eta = etaBin(eta);
+  int bin_pt = ptBin(pt);
+  if (bin_eta == -1 || bin_pt == -1)
+    return 1.;
+  auto n_bins_pt = pt_bins.size();
+  // std::cout << "pt: " << pt << " eta: " << eta << " ptbin: " << bin_pt << " etabin: " << bin_eta << " calib: " <<  calib_factors[(bin_eta*n_bins_pt)+bin_pt] << std::endl;
```
Comment (rekovic) [inline L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:62]: lets leave this commented code as is, useful for checks and validation.
Will clean up in future PRs.

**Comment by rekovic** [comment]
@smuzaffar 
Strange.  I copied your syntax from https://github.com/cms-sw/cmssw/pull/29904#issuecomment-647282374 which worked and the test jobs were successful to here https://github.com/cms-sw/cmssw/pull/30325#issuecomment-647398279
 but here it failed as the path to the data file could not be found.  

Was there some subtlety in the syntax that I missed again ?

**Comment by smuzaffar** [comment]
@rekovic , you used '*' instead of '-' :-)

**Comment by smuzaffar** [comment]
I will update the bot to accept both but for now please use `-`

**Comment by smuzaffar** [comment]
please test

bot should now recognize `*` in test parameters

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc (hunk 1 above)
Comment (kpedro88) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:71]: if there is a reason to keep commented-out code, that reason should be indicated clearly (in a separate comment). In many cases, I think the commented-out code in this PR is not needed. In general, things that are not cleaned up in the initial PR tend to persist forever.

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
urgent

**Comment by kpedro88** [comment]
-1
many comments from #29904 not addressed at all
i'm not going through and repeating all the comments

**Comment by rekovic** [comment]
@kpedro88 
I believe that a huge portion of of them are addressed, and maybe a handful are not. Will go back to #29904 and explicitly mark all the addressed comments.  We can continue the discussion there on the non-addressed comments and address them if needed.

**Comment by rekovic** [comment]
@kpedro88 
My bad.  Yes, indeed there are still many remaining comments.  I thought I had most of them, must have not seen them under the folded part.   Addressing now.

**Comment by rekovic** [comment]
please test

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc (hunk 3 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc]: done

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc (hunk 4 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc]: done

**Comment by rekovic** [comment]
please test

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            bucket->second.push_back(&(*cl3d));
+          }
+        }
+      }
+    }
+  }
+
+  std::sort(selected_multiclusters.begin(), selected_multiclusters.end(), l1tp2::compare_cluster_pt);
+  std::set<const l1t::HGCalMulticluster *> used_clusters;
+  for (const auto cl3d : selected_multiclusters) {
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:111]: `const auto&`
(`const auto` makes a copy, which is slow and unnecessary)

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+      // this is not yet used
+      used_clusters.insert(cl3d);
+      auto eta_phi_bin = get_eta_phi_bin(cl3d);
+
+      for (int eta_bin : {eta_phi_bin.first - 1, eta_phi_bin.first, eta_phi_bin.first + 1}) {
+        for (int phi_bin : {eta_phi_bin.second - 1, eta_phi_bin.second, eta_phi_bin.second + 1}) {
+          auto bucket = etaphi_bins.find(std::make_pair(eta_bin, phi_bin));
+          if (bucket != etaphi_bins.end()) {
+            // this bucket is not empty
+            for (const auto other_cl_ptr : bucket->second) {
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:130]: `const auto&`

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc [hunk 8]
```diff
@@ -0,0 +1,1199 @@
...
+        static float constexpr towerPhiUpperUnitialized = -90;
+        if (l1CaloTower.towerEta() < towerEtaUpperUnitialized && l1CaloTower.towerPhi() < towerPhiUpperUnitialized) {
+          l1CaloTower.setTowerEta(l1t::CaloTools::towerEta(l1CaloTower.towerIEta()));
+          l1CaloTower.setTowerPhi(l1t::CaloTools::towerPhi(l1CaloTower.towerIEta(), l1CaloTower.towerIPhi()));
+        }
+        l1CaloTower.setIsBarrel(true);
+
+        // Add L1EGs if they match in iEta / iPhi
+        // L1EGs are already pT ordered, we will take the ID info for the leading one, but pT as the sum
+        for (auto l1eg : *L1EGXtalClusters) {
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:1145]: this is not the correct fix for the clang warning
it is telling you to use `const auto&` (reference, not copy)

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc (hunk 7 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc]: done.

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc (hunk 8 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc]: done

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc (hunk 6 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc]: done

**Comment by rekovic** [comment]
hmm... not sure why it cannot automatically merge.
The last thing done was force push of the rebased version.

@kpedro88  any idea ?

**Comment by rekovic** [comment]
the problem is with the first commit.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
An example of failing job log messages indicate problems that have nothing to do with this PR [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e8f9f8/7418/runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35/step2_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35.log), but instead with a TrackTrigger PR #30362

I am guessing there is a cross PR dependency in the testing introduced when testing the L1TTrackMatch PR
#30363, which uses both of these 2 PRs.

Try to test alone this PR.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
this looks like a bug in the bot.
Both this PR and PR #30362 (https://github.com/cms-sw/cmssw/pull/30362#issuecomment-649862609)
pointed to the same location of test results 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e8f9f8/7418/summary.html
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e8f9f8/7418/summary.html
atttn @smuzaffar 

Testing alone this PR should pass, I expect.

**Comment by smuzaffar** [comment]
@rekovic , please note that you had started a test to test multiple PRs together ( https://github.com/cms-sw/cmssw/pull/30363#issuecomment-649537943 ) . So when ever a test for PR #30363 will run, it will run for all of these cms-data/L1Trigger-L1TCalorimeter#22,cms-data/L1Trigger-L1TMuon#18,#30325,#30362 and results will points to same location.

**Comment by silviodonato** [comment]
abort

**Comment by silviodonato** [comment]
test parameters
 * pull_request = cms-data/L1Trigger-L1TCalorimeter#22

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
@rekovic I'm sorry I've made a mistake. I was confused from the two tests running at the same time

**Comment by rekovic** [comment]
@silviodonato 
ok.np.

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
@kpedro88 the test of #30325 only was ok https://github.com/cms-sw/cmssw/pull/30325#issuecomment-650095080

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc [hunk 9]
```diff
@@ -0,0 +1,1199 @@
...
+    float et = decoder_->hcaletValue(hit.id(), hit.t0());
+    if (et <= 0)
+      continue;
+    if (!(hcTopology_->validHT(hit.id()))) {
+      LogError("L1EGCrystalClusterEmulatorProducer")
+          << " -- Hcal hit DetID not present in HCAL Geom: " << hit.id() << std::endl;
+      throw cms::Exception("L1EGCrystalClusterEmulatorProducer");
+      continue;
+    }
+    std::vector<HcalDetId> hcId = theTrigTowerGeometry.detIds(hit.id());
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:403]: repeated comment:
`const std::vector<HcalDetId>& hcId = theTrigTowerGeometry.detIds(hit.id());`

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc [hunk 10]
```diff
@@ -0,0 +1,1199 @@
...
+  //*********************************************************
+  //******************** Do layer 2 *************************
+  //*********************************************************
+
+  // Definition of L2 outputs
+  float HCAL_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];  // 3 L2 cards send each 48 links with 17 towers
+  float ECAL_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];  // 3 L2 cards send each 48 links with 17 towers
+  int iEta_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];    // 3 L2 cards send each 48 links with 17 towers
+  int iPhi_tower_L2Card[n_links_GCTcard][n_towers_cardEta][3];    // 3 L2 cards send each 48 links with 17 towers
+  float energy_cluster_L2Card[n_links_GCTcard][2][3];             // 3 L2 cards send each 48 links with 2 clusters
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:767]: repeated comments:
magic numbers 2,3 should be named constants (throughout this section of code)
also, eliminate duplicate comments (just include each comment once)

File: L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
+import FWCore.ParameterSet.Config as cms
+
+from Configuration.StandardSequences.Eras import eras
+
+process = cms.Process("L1AlgoTest",eras.Phase2_trigger)
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py:5]: repeated comment:

please use the actual era for the phase 2 workflow being tested
(phase2_trigger is an internal use modifier, Phase2_trigger with that capitalization is not anything)

"please use the actual era for the phase 2 workflow being tested" does not mean "use the Phase2 era"

different scenarios have different associated eras, seen in https://github.com/cms-sw/cmssw/blob/f18f5a71e7278be55d457d17e38379b11cd04c19/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py#L740

File: L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+#include <iostream>
+
+namespace l1tp2 {
+  class ParametricCalibration {
+  public:
+    ParametricCalibration() {}
+    ParametricCalibration(const edm::ParameterSet& cpset);
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+    float operator()(const float pt, const float abseta) const {
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h:18]: multiline implementations should typically be put in the .cc file
(unless e.g. it is desired to be inlined)

**Comment by kpedro88** [comment]
in addition to the few still unaddressed comments I've highlighted here, the code remains littered with magic numbers. But apparently everyone else is satisfied with allowing possible inconsistencies that could invalidate the results from O(100K) CPU hours spent for a full production. Oh well.

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc (hunk 10 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:767]: done

File: L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py (hunk 11 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py:5]: done.

File: L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h (hunk 12 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h]: done

File: L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc (hunk 9 above)
Comment (rekovic) [reply L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc]: done

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade
(needs to be added to cms-bot package categories)

**Comment by silviodonato** [comment]
Thank you for the reminder. @smuzaffar  @mrodozov could you assign it to l1 and upgrade?

**Comment by silviodonato** [comment]
cms-bot package: https://github.com/cms-sw/cms-bot/pull/1336
cms-dist: https://github.com/cms-sw/cmsdist/pull/6004
(https://github.com/cms-data/L1Trigger-L1TCalorimeter/pull/22)

**Comment by silviodonato** [comment]
merge

**Pull request merged**